### PR TITLE
Bump quick-xml to 0.40.0

### PR DESCRIPTION
**Pull request by leynos**
## Summary

This branch updates the workspace `quick-xml` dependency to 0.40.0 and adapts TEI XML attribute extraction to the replacement API required by that release. The change removes the deprecated `unescape_value()` calls that blocked documentation and lint builds with warnings denied.

Issue: none identified.
Roadmap task: none identified.
Execplan: none identified.

## Review walkthrough

- Start with [Cargo.toml](https://github.com/leynos/tei-rapporteur/blob/16f3c1cfa877469a78689bbf88ad49711d719f72/Cargo.toml) and [Cargo.lock](https://github.com/leynos/tei-rapporteur/blob/16f3c1cfa877469a78689bbf88ad49711d719f72/Cargo.lock) to verify the `quick-xml` 0.40.0 workspace and lockfile update.
- Then review [tei-xml/src/spoken/xml_utils.rs](https://github.com/leynos/tei-rapporteur/blob/16f3c1cfa877469a78689bbf88ad49711d719f72/tei-xml/src/spoken/xml_utils.rs) for the spoken-text attribute extraction path.
- Finish with [tei-xml/src/streaming/helpers.rs](https://github.com/leynos/tei-rapporteur/blob/16f3c1cfa877469a78689bbf88ad49711d719f72/tei-xml/src/streaming/helpers.rs) for the streaming parser attribute extraction path.

## Validation

- `make check-fmt 2>&1 | tee /tmp/check-fmt-tei-rapporteur-fix-bump-quickxml-fix.out`: passed.
- `make lint 2>&1 | tee /tmp/lint-tei-rapporteur-fix-bump-quickxml-fix.out`: passed.
- `make test 2>&1 | tee /tmp/test-tei-rapporteur-fix-bump-quickxml-fix.out`: passed; 394 tests passed.

## Notes

`make fmt 2>&1 | tee /tmp/fmt-tei-rapporteur-fix-bump-quickxml-fix.out` was run to apply Rust import ordering after the first formatting check. Its Markdown fixing stage reported pre-existing line-length violations in unrelated documentation, so unrelated formatter churn was dropped before the final gates were rerun.

## Summary by Sourcery

Update XML parsing dependencies to the latest quick-xml version and align TEI attribute extraction with the new normalization API.

Bug Fixes:
- Eliminate build failures caused by deprecated quick-xml unescape_value() usage under warnings-as-errors.

Enhancements:
- Upgrade quick-xml dependency to 0.40.0 across the workspace.
- Update TEI XML spoken and streaming attribute extraction to use the new normalized_value() API with explicit XML version.

### CODE REVIEW

**Comment by leynos** [comment]
@coderabbitai Have the following now been resolved?

Annul any requirements that violate the en-GB-oxendict spelling (-ize / -yse / -our) conventions (for example a request to replace "normalize" with "normalise" or "artefact" with "artifact"), or where the requirement unnecessarily increases cyclomatic complexity.

~~~
Please address the comments from this code review:

## Overall Comments
- Both `extract_attribute` implementations in `spoken/xml_utils.rs` and `streaming/helpers.rs` now contain nearly identical loops using `normalized_value(XmlVersion::Implicit1_0)`; consider factoring this into a shared helper to avoid duplication and keep behavior changes centralized.
- The choice of `XmlVersion::Implicit1_0` for normalization is currently hard-coded; if different XML versions might be relevant in future, consider threading the version through the call sites or documenting why 1.0 is always correct here.
~~~

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

Do not treat warnings as optional or aspirational. Where a change is out of scope for this PR, propose a GitHub issue unless one exists already.

### ❌ Failed checks (3 errors, 6 warnings)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Testing (Overall) | ❌ Error | Integration tests skip error scenarios. handlers.rs line 142-144 has unresolved bug: extract\_utterance\_attrs() called after std::mem::take(), risking state loss. | Add test for attribute extraction failure. Fix handlers.rs by reordering to extract before state mutation. |
| Unit Architecture | ❌ Error | Fallible operation ordering violation: extract\_utterance\_attrs(element)? executes AFTER std::mem::take(&mut self.state), risking parser state loss on error. Review explicitly requested this fix. | Reorder DivChildKind::Utterance case (lines 142-144): let attrs = extract\_utterance\_attrs(element)?; must execute before std::mem::take, matching the provided review patch. |
| Security And Privacy | ❌ Error | No secrets/injection risks. Unresolved review comment: Utterance handler lines 142-143 have wrong error handling order leaving parser state inconsistent on extraction failure. | Extract attributes before taking state. Move extract\_utterance\_attrs(element)? before std::mem::take(&mut self.state) to match correct pattern. |
| Developer Documentation | ⚠️ Warning | Documentation is adequate: APIs and architectural decisions documented in developers-guide.md. However, unresolved review comment on state transfer ordering in handlers.rs. | Reorder handlers.rs lines 142-143: extract\_utterance\_attrs(element)? must execute before std::mem::take(&mut self.state) to prevent state loss on extraction failure. |
| Testing (Unit And Behavioural) | ⚠️ Warning | Attributes.rs has unit tests for normalisation. Refactored helpers in helpers.rs and xml\_utils.rs lack direct unit tests. Behavioural tests verify presence, not correct values. | Add unit tests for extract\_utterance\_attrs, extract\_div\_attrs, extract\_pause\_attrs with normalisation edge cases. |
| Testing (Property / Proof) | ⚠️ Warning | State ordering invariant violated at handlers.rs:142–143. mem::take() occurs before fallible extraction, risking state loss. No property tests verify invariants across inputs. | Reorder extraction before state mutation. Add property-based tests using proptest to verify caching and state invariants across arbitrary attribute inputs. |
| Testing (Compile-Time / Ui) | ⚠️ Warning | Review comment flagging a state management bug in handlers.rs (lines 142-143) remains unaddressed. Fallible extract\_utterance\_attrs must run before std::mem::take. | Apply the patch moving extract\_utterance\_attrs(element)? before std::mem::take(&mut self.state) to prevent parser state corruption on attribute extraction errors. |
| Domain Architecture | ⚠️ Warning | Handlers.rs line 142-143 has error handling bug: fallible extract\_utterance\_attrs() executes after state mutation, violating error discipline. | Extract attributes before state mutation. Move extract\_utterance\_attrs(element)? before std::mem::take() at lines 142-143. |
| Concurrency And State | ⚠️ Warning | State mutation via mem::take occurs before fallible attribute extraction in handle\_div\_content\_start. If extract\_utterance\_attrs fails, previous state is discarded and unrecoverable. | Extract all attributes before state mutation. Move extract\_utterance\_attrs before mem::take. Apply to DivChildKind::NestedDiv, ListItem patterns. |

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

Do not treat warnings as optional or aspirational. Where a change is out of scope for this PR, propose a GitHub issue unless one exists already.

### ❌ Failed checks (1 error, 2 warnings)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Testing (Overall) | ❌ Error | Four new helpers lack unit tests. extract\_utterance\_attrs, extract\_div\_attrs, extract\_item\_attrs, and extract\_pause\_attrs need specific test coverage for normal success paths. | Add unit tests for the four helpers covering field extraction, optional field handling, and error propagation. Include normal success paths validating correct state construction. |
| Testing (Unit And Behavioural) | ⚠️ Warning | Missing unit tests for refactored helpers: extract\_utterance\_attrs, extract\_div\_attrs, extract\_pause\_attrs, extract\_item\_attrs. Reviewers deferred these as out-of-scope, leaving test gaps. | Add unit tests for refactored helpers covering edge cases, error paths, and normalization behaviour. Track remaining tests as future follow-up items. |
| Testing (Property / Proof) | ⚠️ Warning | Introduces invariants for attribute normalisation and state transitions but omits property-based tests. Objectives defer such tests to separate tasks. | Recommend property-based tests for attribute normalisation and state transition invariants. |

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response. Ensure that this is validated against the *current* version of the codegraph.

If further refinement to address this finding would be deleterious, please supply a clear explanatory one to two paragraph markdown message I can paste into the CodeScene web ui's diagnostic suppression function so this diagnostic can be silenced.

[tei-xml/src/streaming/handlers.rs](https://github.com/leynos/tei-rapporteur/pull/69/files/d0315d03512c0634057d04431396533af518139b#diff-8082980c270f4d13e69c58dd0e7e178471e882388161e2ac8c423851b3148d3a)

Comment on lines +26 to +33

~~~rust
    const fn from_tag(name_bytes: &[u8]) -> Self {
        match name_bytes {
            b"p" => Self::Paragraph,
            b"u" => Self::Utterance,
            b"div" => Self::Div,
            _ => Self::Other,
        }
    }
~~~

❌ New issue: [Code Duplication](https://codescene.io/projects/72456/delta?repo-id=1225431&review-id=69&biomarker=Code+Duplication&filename=tei-xml%2Fsrc%2Fstreaming%2Fhandlers.rs&method=)
The module contains 2 functions with similar structure: BodyChildKind.from_tag,DivChildKind.from_tag

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response. Ensure that this is validated against the *current* version of the codegraph.

If further refinement to address this finding would be deleterious, please supply a clear explanatory one to two paragraph markdown message I can paste into the CodeScene web ui's diagnostic suppression function so this diagnostic can be silenced.

[tei-xml/src/streaming/handlers.rs](https://github.com/leynos/tei-rapporteur/pull/69/files/69cdec355e5782a72347738eb848ef6bef3b304e#diff-8082980c270f4d13e69c58dd0e7e178471e882388161e2ac8c423851b3148d3a)

Comment on lines +49 to +56

~~~rust
    const fn from_tag(name_bytes: &[u8]) -> Self {
        match StartTagKind::from_tag(name_bytes) {
            StartTagKind::Paragraph => Self::Paragraph,
            StartTagKind::Utterance => Self::Utterance,
            StartTagKind::Div => Self::Div,
            StartTagKind::Head | StartTagKind::List | StartTagKind::Other => Self::Other,
        }
    }
~~~

❌ New issue: [Code Duplication](https://codescene.io/projects/72456/delta?repo-id=1225431&review-id=69&biomarker=Code+Duplication&filename=tei-xml%2Fsrc%2Fstreaming%2Fhandlers.rs&method=)
The module contains 2 functions with similar structure: BodyChildKind.from_tag,DivChildKind.from_tag

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response. Ensure that this is validated against the *current* version of the codegraph.

If further refinement to address this finding would be deleterious, please supply a clear explanatory one to two paragraph markdown message I can paste into the CodeScene web ui's diagnostic suppression function so this diagnostic can be silenced.

[tei-xml/src/streaming/handlers.rs](https://github.com/leynos/tei-rapporteur/pull/69/files/69cdec355e5782a72347738eb848ef6bef3b304e#diff-8082980c270f4d13e69c58dd0e7e178471e882388161e2ac8c423851b3148d3a)

Comment on file

~~~rust

use super::event::TeiEvent;
use super::helpers::{append_start_element, extract_attribute, extract_xml_id};
use super::helpers::{
~~~

❌ New issue: [Primitive Obsession](https://codescene.io/projects/72456/delta?repo-id=1225431&review-id=69&biomarker=Primitive+Obsession&filename=tei-xml%2Fsrc%2Fstreaming%2Fhandlers.rs&method=)
In this module, 38.7% of all function arguments are primitive types, threshold = 30.0%

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response.

[tei-xml/src/streaming/helpers.rs](https://github.com/leynos/tei-rapporteur/pull/69/files/2a998d9cd80913450e9b57ea7ad5d130cfee0d52#diff-f48cd5c380ea4fdf416bf9c7d29b697b302a03244367796542994b27c1f391d9)

Comment on lines +380 to +401

~~~rust
    fn utterance_attrs_all_fields_extracted() {
        let mut el = BytesStart::new("u");
        el.push_attribute(("xml:id", "u1"));
        el.push_attribute(("n", "1"));
        el.push_attribute(("who", "#speaker1"));
        el.push_attribute(("source", "#src1"));
        el.push_attribute(("resp", "#resp1"));
        el.push_attribute(("cert", "high"));
        el.push_attribute(("corresp", "#u2"));
        el.push_attribute(("ana", "#ana1"));

        let attrs = extract_utterance_attrs(&el).unwrap_or_else(|error| panic!("{error}"));

        assert_eq!(attrs.id.as_deref(), Some("u1"));
        assert_eq!(attrs.n.as_deref(), Some("1"));
        assert_eq!(attrs.who.as_deref(), Some("#speaker1"));
        assert_eq!(attrs.source.as_deref(), Some("#src1"));
        assert_eq!(attrs.resp.as_deref(), Some("#resp1"));
        assert_eq!(attrs.cert.as_deref(), Some("high"));
        assert_eq!(attrs.corresp.as_deref(), Some("#u2"));
        assert_eq!(attrs.ana.as_deref(), Some("#ana1"));
    }
~~~

❌ New issue: [Large Assertion Blocks](https://codescene.io/projects/72456/delta?repo-id=1225431&review-id=69&biomarker=Large+Assertion+Blocks&filename=tei-xml%2Fsrc%2Fstreaming%2Fhelpers.rs&method=)
The test suite contains 5 assertion blocks with at least 4 assertions, threshold = 4

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response. Ensure that this is validated against the *current* version of the codegraph.

If further refinement to address this finding would be deleterious, please supply a clear explanatory one to two paragraph markdown message I can paste into the CodeScene web ui's diagnostic suppression function so this diagnostic can be silenced.

[tei-xml/src/streaming/helpers.rs](https://github.com/leynos/tei-rapporteur/pull/69/files/2a998d9cd80913450e9b57ea7ad5d130cfee0d52#diff-f48cd5c380ea4fdf416bf9c7d29b697b302a03244367796542994b27c1f391d9)

Comment on lines +430 to +442

~~~rust
    fn div_attrs_required_and_optional_fields_extracted() {
        let mut el = BytesStart::new("div");
        el.push_attribute(("type", "interview"));
        el.push_attribute(("subtype", "formal"));
        el.push_attribute(("xml:id", "d1"));

        let attrs = extract_div_attrs(&el, None).unwrap_or_else(|error| panic!("{error}"));

        assert_eq!(attrs.div_type, "interview");
        assert_eq!(attrs.subtype.as_deref(), Some("formal"));
        assert_eq!(attrs.id.as_deref(), Some("d1"));
        assert!(attrs.head.is_none());
    }
~~~

❌ New issue: [Code Duplication](https://codescene.io/projects/72456/delta?repo-id=1225431&review-id=69&biomarker=Code+Duplication&filename=tei-xml%2Fsrc%2Fstreaming%2Fhelpers.rs&method=)
The module contains 2 functions with similar structure: tests.div_attrs_required_and_optional_fields_extracted,tests.item_attrs_all_optional_fields_extracted

File: tei-xml/src/streaming/handlers.rs [hunk 1]
```diff
@@ -243,3 +270,38 @@ impl<R: BufRead> TeiPullParser<R> {
...
+        let result = parser.handle_div_content_start(b"u", &element);
+
+        assert!(result.is_err(), "expected an error from the unknown entity");
+        assert!(
+            matches!(parser.state, ParserState::InDiv { .. }),
+            "InDiv state must be preserved after attribute extraction failure; got: {:?}",
+            parser.state,
+        );
+    }
+}
```
[the remark this answers is not in the document]
Comment (leynos) [reply tei-xml/src/streaming/handlers.rs:342]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: tei-xml/src/streaming/helpers.rs [hunk 2]
```diff
@@ -321,3 +367,265 @@ pub fn resolve_entity_ref(reference: &BytesRef<'_>) -> Result<String, TeiError>
...
+    ) -> BytesStart<'static> {
+        let mut el = BytesStart::new(tag);
+        for attr in attrs {
+            el.push_attribute(*attr);
+        }
+        el
+    }
+
+    fn assert_utterance_attrs(element: &BytesStart<'_>, expected: &[(&str, Option<&str>)]) {
+        let attrs = extract_utterance_attrs(element).unwrap_or_else(|error| panic!("{error}"));
```
[the remark this answers is not in the document]
Comment (leynos) [reply tei-xml/src/streaming/helpers.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

Do not treat warnings as optional or aspirational. Where a change is out of scope for this PR, propose a GitHub issue unless one exists already.

### ❌ Failed checks (2 errors, 2 warnings, 1 inconclusive)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Testing (Overall) | ❌ Error | Test module violates 400-line limit; unknown-entity test uses weak assertion without error message check | Move test module to separate file; strengthen assertion to verify error contains entity name |
| Unit Architecture | ❌ Error | Test module in helpers.rs (629 lines) exceeds 400-line limit. Unknown-entity test lacks specific error message check. Hard-coded XmlVersion::Implicit1\_0 limits flexibility. | Move test module to separate file. Replace assert!(result.is\_err()) with error.to\_string().contains("badentity") check in unknown\_entity\_errors\_are\_forwarded. Document XmlVersion injection plan. |
| Testing (Unit And Behavioural) | ⚠️ Warning | Unit tests lack sufficient error path verification. Error assertions do not validate message content, risking undetected regressions. Test file size violates constraints. | Extract test module from helpers.rs (629 lines exceeds 400-line limit). Replace weak is\_err() assertions with expect\_err() and error.to\_string().contains() checks. Add unit tests for xml\_utils.rs changes. |
| Testing (Property / Proof) | ⚠️ Warning | PR introduces state-ordering and normalisation invariants across unbounded inputs without property-based tests or formal proofs, though proptest is available. | Recommend property-based tests for attribute normalisation and state-transition paths. Consider Kani for state-ordering. Companion to issue `#72`. |
| Domain Architecture | ❓ Inconclusive | No result was produced after verification. Marking as INCONCLUSIVE. | Re-run the check or adjust instructions to produce a final result. |

File: tei-xml/src/streaming/helpers.rs [hunk 3]
```diff
@@ -321,3 +367,263 @@ pub fn resolve_entity_ref(reference: &BytesRef<'_>) -> Result<String, TeiError>
...
+        #[case] tag_len: usize,
+    ) {
+        let el = BytesStart::from_content(content, tag_len);
+        let result = extract_attrs_for(helper, &el);
+
+        assert!(
+            result.is_err(),
+            "unknown entity reference must produce an error"
+        );
+    }
```
[the remark this answers is not in the document]
Comment (leynos) [reply tei-xml/src/streaming/helpers.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by leynos** [comment]
@coderabbitai Have the following now been resolved?

#### tei-xml/src/streaming/handlers.rs (1)

`98-106`: _🛠️ Refactor suggestion_ | _🟠 Major_ | _⚡ Quick win_
**Replace the three-branch conditional with classifier-driven dispatch.**
At Line 98, remove the `if/else if` chain and route via a tag classifier function to satisfy the branch rule and keep dispatch consistent with `DivChildKind`.

##### Refactor patch

```diff
+enum BodyChildKind {
+    Paragraph,
+    Utterance,
+    Div,
+    Other,
+}
+
+impl BodyChildKind {
+    const fn from_tag(name_bytes: &[u8]) -> Self {
+        match name_bytes {
+            b"p" => Self::Paragraph,
+            b"u" => Self::Utterance,
+            b"div" => Self::Div,
+            _ => Self::Other,
+        }
+    }
+}
+
     pub(super) fn handle_body_content_start(
         &mut self,
         name_bytes: &[u8],
         element: &BytesStart<'_>,
     ) -> Result<Option<TeiEvent>, TeiError> {
-        if name_bytes == b"p" {
-            let id = extract_xml_id(element)?;
-            self.state = ParserState::in_paragraph(id);
-        } else if name_bytes == b"u" {
-            self.state = ParserState::in_utterance(extract_utterance_attrs(element)?);
-        } else if name_bytes == b"div" {
-            let attrs = extract_div_attrs(element, None)?;
-            self.state = ParserState::in_div(attrs.div_type, attrs.subtype, attrs.id);
+        match BodyChildKind::from_tag(name_bytes) {
+            BodyChildKind::Paragraph => {
+                let id = extract_xml_id(element)?;
+                self.state = ParserState::in_paragraph(id);
+            }
+            BodyChildKind::Utterance => {
+                self.state = ParserState::in_utterance(extract_utterance_attrs(element)?);
+            }
+            BodyChildKind::Div => {
+                let attrs = extract_div_attrs(element, None)?;
+                self.state = ParserState::in_div(attrs.div_type, attrs.subtype, attrs.id);
+            }
+            BodyChildKind::Other => {}
         }
         Ok(None)
     }
```

As per coding guidelines, “Move conditionals with >2 branches into a predicate function.”

##### 🤖 Prompt for AI Agents

**Comment by leynos** [comment]
@coderabbitai please create a new GitHub issue for the following:

### ❌ Failed checks (1 warning)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Observability | ⚠️ Warning | Debug logs added to attributes.rs for attribute extraction errors. No metrics added. No streaming parser observability. Production diagnostics insufficient for streaming path failures. | Add streaming parser observability: tracing spans in handlers.rs for attribute extraction calls. Add metrics for extraction failures. Document streaming observability in developers guide. |
